json-endpoints/jetpack/class.jetpack-json-api-cron-endpoint.php

```diff
@@ -0,0 +1,228 @@
+	}
+}
+
+// POST /sites/%s/cron/schedule
+class Jetpack_JSON_API_Cron_Schedule_Endpoint extends Jetpack_JSON_API_Cron_Endpoint {
+
+	protected function result() {
+		$args = $this->input();
+		if ( ! isset( $args['timestamp'] ) ) {
+			return new WP_Error( 'missing_argument', 'Please provide the timestamp argument' );
+		}
+
+		if ( ! is_int( $args['timestamp'] ) || $args['timestamp'] < time() ) {
+			return new WP_Error( 'timestamp-invalid', 'Please provide timestamp that is an integer and set in the future' );
+		}
+
+		if ( ! isset( $args['hook'] ) ) {
+			return new WP_Error( 'missing_argument', 'Please provide the hook argument' );
+		}
+
+		$hook = $this->sanitize_hook( $args['hook'] );
+
+		$locked = $this->is_cron_locked( microtime( true ) );
+		if ( is_wp_error( $locked ) ) {
+			return $locked;
+		}
+
+		$arguments = $this->resolve_arguments();
+
+		if ( wp_next_scheduled( $hook, $arguments ) ) {
+			return new WP_Error( 'event-already-scheduled', 'This event is ready scheduled' );
+		}
+
+		if ( ! isset( $args['recurrence'] ) ) {
+			$lock = $this->lock_cron();
+			wp_schedule_single_event( $args['timestamp'], $hook, $arguments );
+			$this->maybe_unlock_cron( $lock );
+			return array( 'success' => true );
+		}
+		
+		$schedules = wp_get_schedules();
+		if ( ! isset( $schedules[ $args['recurrence'] ] ) ) {
+			return new WP_Error( 'invalid-recurrence', 'Please provide a valid recurrence argument' );
+		}
+		
+		$lock = $this->lock_cron();
+		wp_schedule_event( $args['timestamp'], $args['recurrence'], $hook, $arguments );
+		$this->maybe_unlock_cron( $lock );
+
+		return array( 'success' => true );
+	}
+}
+
+// POST /sites/%s/cron/unschedule
+class Jetpack_JSON_API_Cron_Unschedule_Endpoint extends Jetpack_JSON_API_Cron_Endpoint {
+
+	protected function result() {
+		$args = $this->input();
+
+		if ( !isset( $args['hook'] ) ) {
+			return new WP_Error( 'missing_argument', 'Please provide the hook argument' );
+		}
+
+		$hook = $this->sanitize_hook( $args['hook'] );
+
+		$locked = $this->is_cron_locked( microtime( true ) );
+		if ( is_wp_error( $locked ) ) {
+			return $locked;
+		}
+
+		$crons = _get_cron_array();
+		if ( empty( $crons ) ) {
+			return new WP_Error( 'cron-not-present', 'Unable to unschedule an event, no events in the cron' );
+		}
+
+		$arguments = $this->resolve_arguments();
+
+		if ( isset( $args['timestamp'] ) ) {
+			$prevent_timestamp = wp_next_scheduled( $hook, $arguments );
+			if ( $args['timestamp'] != $prevent_timestamp ) {
+				return new WP_Error( 'event-not-present', 'Unable to unschedule the event, the event doesn\'t exist' );
+			}
+
+			$lock = $this->lock_cron();
+			wp_unschedule_event( $args['timestamp'], $hook, $arguments );
+			$this->maybe_unlock_cron( $lock );
+			return array( 'success' => true );
+		}
+		$lock = $this->lock_cron();
+		wp_clear_scheduled_hook( $hook, $arguments );
+		$this->maybe_unlock_cron( $lock );
+		return array( 'success' => true );
+	}
+}
```

Comment:
I believe we're supposed to end files with an empty newline. It's in the Calypso guidelines. 

I've never truly understood why, but here is a case for it. http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline